File: APP/lib/models/hall.dart
import 'package:json_annotation/json_annotation.dart';

part 'hall.g.dart';

enum HallType { male, female }
enum BedNumberingFormat { standard, custom }
enum BedStatus { vacant, occupied, reserved, maintenance }

@JsonSerializable()
class Hall {
  final String id;
  final String name;
  final String code;
  final HallType type;
  final int capacity;
  final int currentOccupancy;
  final int availableBeds;
  final int specialNeedsOccupancy;
  final List<Bed> beds;
  final BedNumberingFormat numberingFormat;
  final BedNumberingConfig? numberingConfig;
  final DateTime createdAt;
  final DateTime updatedAt;

  Hall({
    required this.id,
    required this.name,
    required this.code,
    required this.type,
    required this.capacity,
    required this.currentOccupancy,
    required this.availableBeds,
    required this.specialNeedsOccupancy,
    required this.beds,
    required this.numberingFormat,
    this.numberingConfig,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Hall.fromJson(Map<String, dynamic> json) => _$HallFromJson(json);
  Map<String, dynamic> toJson() => _$HallToJson(this);
}

@JsonSerializable()
class Bed {
  final String id;
  final String number;
  final String hallId;
  final String hallCode;
  final BedStatus status;
  final String? pilgrimId;
  final bool isSpecialNeeds;
  final bool? isDoubleBed;
  final String? companionBedId;
  final DateTime? lastAssignedAt;
  final DateTime? lastVacatedAt;
  final String? maintenanceNotes;

  Bed({
    required this.id,
    required this.number,
    required this.hallId,
    required this.hallCode,
    required this.status,
    this.pilgrimId,
    required this.isSpecialNeeds,
    this.isDoubleBed,
    this.companionBedId,
    this.lastAssignedAt,
    this.lastVacatedAt,
    this.maintenanceNotes,
  });

  factory Bed.fromJson(Map<String, dynamic> json) => _$BedFromJson(json);
  Map<String, dynamic> toJson() => _$BedToJson(this);
}

@JsonSerializable()
class BedNumberingConfig {
  final String? prefix;
  final String? suffix;
  final int startNumber;
  final int padding;
  final String separator;

  BedNumberingConfig({
    this.prefix,
    this.suffix,
    required this.startNumber,
    required this.padding,
    required this.separator,
  });

  factory BedNumberingConfig.fromJson(Map<String, dynamic> json) => _$BedNumberingConfigFromJson(json);
  Map<String, dynamic> toJson() => _$BedNumberingConfigToJson(this);
}

File: frontend/types/hall.ts
import { BedStatus } from './pilgrim';

export type HallType = 'male' | 'female';
export type BedNumberingFormat = 'standard' | 'custom';

export interface Hall {
  id: string;
  name: string;
  code: string;
  type: HallType;
  capacity: number;
  currentOccupancy: number;
  availableBeds: number;
  specialNeedsOccupancy: number;
  beds: Bed[];
  numberingFormat: BedNumberingFormat;
  numberingConfig?: BedNumberingConfig;
  createdAt: Date;
  updatedAt: Date;
}

export interface Bed {
  id: string;
  number: string;
  hallId: string;
  hallCode: string;
  status: BedStatus;
  pilgrimId?: string;
  pilgrimName?: string;
  isSpecialNeeds: boolean;
  isDoubleBed?: boolean;
  companionBedId?: string;
  lastAssignedAt?: Date;
  lastVacatedAt?: Date;
  maintenanceNotes?: string;
}

export interface BedNumberingConfig {
  prefix?: string;
  suffix?: string;
  startNumber: number;
  padding: number;
  separator: string;
}

export interface HallStatistics {
  totalHalls: number;
  totalBeds: number;
  totalOccupied: number;
  totalAvailable: number;
  totalMaintenance: number;
  totalReserved: number;
  occupancyRate: number;
  maleHalls: {
    count: number;
    beds: number;
    occupied: number;
    occupancyRate: number;
  };
  femaleHalls: {
    count: number;
    beds: number;
    occupied: number;
    occupancyRate: number;
  };
  specialNeedsOccupied: number;
  byHall: Record<string, {
    name: string;
    occupancy: number;
    capacity: number;
    rate: number;
  }>;
}

export interface CreateHallDto {
  name: string;
  code: string;
  type: HallType;
  capacity: number;
  numberingFormat?: BedNumberingFormat;
  numberingConfig?: BedNumberingConfig;
}

export interface UpdateHallDto extends Partial<CreateHallDto> {
  id: string;
}

export interface HallFilters {
  search?: string;
  type?: HallType;
  minOccupancy?: number;
  maxOccupancy?: number;
  hasAvailableBeds?: boolean;
  hasSpecialNeeds?: boolean;
}

export interface BedFilters {
  status?: BedStatus[];
  isSpecialNeeds?: boolean;
  search?: string;
}

export interface BedAssignmentDto {
  pilgrimId: string;
  bedId: string;
  isSpecialNeeds?: boolean;
  companionBedId?: string;
}

export interface BedTransferDto {
  pilgrimId: string;
  fromBedId: string;
  toBedId: string;
  reason?: string;
}

export interface HallOccupancyTrend {
  date: Date;
  hallId: string;
  occupancy: number;
  arrivals: number;
  departures: number;
}

export const DEFAULT_NUMBERING_CONFIG: Record<HallType, BedNumberingConfig> = {
  male: {
    separator: '-',
    startNumber: 100,
    padding: 0,
  },
  female: {
    prefix: '',
    separator: '-',
    startNumber: 1,
    padding: 3,
  }
};

export const HALL_CODE_PATTERN = /^[A-Z]$/;
export const MAX_HALL_CAPACITY = 200;
export const MIN_HALL_CAPACITY = 10;

export interface PaginatedHallsResponse<T> {
  data: T[];
  total: number;
  page: number;
  limit: number;
  totalPages: number;
  hasNext: boolean;
  hasPrevious: boolean;
}